### PR TITLE
[RSDK-4110] Feed IMU data to cartographer

### PR DESCRIPTION
**Pull request by PStrutz**
Create second sensorprocess thread for IMU
Create addIMUReading function in cartofacade and C code
Create CAPI wrapper for addIMUReading

TESTING
*note that this PR will not yet contain the offline tests, as there is a different ticket for creating the replay IMU stub dependencies, upon which the tests are dependent

### CODE REVIEW

File: cartofacade/capi.go [hunk 1]
```diff
@@ -74,6 +75,16 @@ type GetPosition struct {
 	ComponentReference string
 }
 
+// IMUReading holds values for linear acceleration and angular velocity to be converted into c
+type IMUReading struct {
+	LinAccX float64
```
Comment (PStrutz) [inline cartofacade/capi.go:80]: Let's make these r3 Vectors

File: cartofacade/capi.go [hunk 2]
```diff
@@ -453,6 +501,12 @@ func toError(status C.int) error {
 		return errors.New("VIAM_CARTO_NOT_IN_STARTED_STATE")
 	case C.VIAM_CARTO_NOT_IN_TERMINATABLE_STATE:
 		return errors.New("VIAM_CARTO_NOT_IN_TERMINATABLE_STATE")
+	case C.VIAM_CARTO_IMU_CONFIG_INVALID:
+		return errors.New("VIAM_CARTO_IMU_CONFIG_INVALID")
+	case C.VIAM_CARTO_IMU_READING_EMPTY:
```
Comment (PStrutz) [inline cartofacade/capi.go:506]: Remove this

File: cartofacade/capi.go [hunk 3]
```diff
@@ -453,6 +501,12 @@ func toError(status C.int) error {
 		return errors.New("VIAM_CARTO_NOT_IN_STARTED_STATE")
 	case C.VIAM_CARTO_NOT_IN_TERMINATABLE_STATE:
 		return errors.New("VIAM_CARTO_NOT_IN_TERMINATABLE_STATE")
+	case C.VIAM_CARTO_IMU_CONFIG_INVALID:
```
Comment (PStrutz) [inline cartofacade/capi.go:504]: rename to ...ENABLED_INVALID

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 4]
```diff
@@ -345,7 +346,7 @@ void CartoFacade::IOInit() {
 
     {
         std::lock_guard<std::mutex> lk(map_builder_mutex);
-        map_builder.StartLidarTrajectoryBuilder();
+        map_builder.StartTrajectoryBuilder(algo_config.use_imu_data);
```
Comment (PStrutz) [inline viam-cartographer/src/carto_facade/carto_facade.cc:349]: Are both call necessary?

File: cartofacade/carto_facade_test.go
Comment (PStrutz) [inline cartofacade/carto_facade_test.go:1]: Will be adding high-level description of testing Friday morning

File: sensorprocess/sensorprocess_test.go
Comment (PStrutz) [inline sensorprocess/sensorprocess_test.go:1]: Will be adding high-level description of testing Friday morning

File: cartofacade/capi_test.go
Comment (PStrutz) [inline cartofacade/capi_test.go:1]: Will be adding high-level description of testing Friday morning

File: cartofacade/capi_mock.go
Comment (PStrutz) [inline cartofacade/capi_mock.go:1]: Will be adding high-level description of testing Friday morning

File: cartofacade/carto_facade_test.go
Comment (PStrutz) [inline cartofacade/carto_facade_test.go:1]: Will be adding high-level description of testing Friday morning

File: cartofacade/carto_facade_test.go [hunk 5]
```diff
@@ -331,13 +331,13 @@ func TestAddSensorReading(t *testing.T) {
```
Comment (jeremyrhyde) [inline cartofacade/carto_facade_test.go:331]: [note] Testing to be added to cart_facade_test.go

File: cartofacade/capi_test.go [hunk 6]
```diff
@@ -49,7 +49,7 @@ func testAddSensorReading(t *testing.T, vc Carto, pcdPath string, timestamp time
 	err = pointcloud.ToPCD(pc, buf, pcdType)
```
Comment (jeremyrhyde) [inline cartofacade/capi_test.go:49]: [note] Testing to be added to capi_test.go

File: sensorprocess/sensorprocess_test.go [hunk 7]
```diff
@@ -47,8 +47,8 @@ func TestAddSensorReadingOffline(t *testing.T) {
 		LidarDataRateMsec: 200,
```
Comment (jeremyrhyde) [inline sensorprocess/sensorprocess_test.go:47]: [note] Testing to be added to sensorprocess_test.go

File: sensorprocess/sensorprocess.go [hunk 8]
```diff
@@ -109,3 +112,96 @@ func tryAddSensorReading(ctx context.Context, reading []byte, readingTime time.T
...
+func addIMUReading(
+	ctx context.Context,
+	config Config,
+) bool {
+	tsr, err := config.IMU.TimedIMUSensorReading(ctx)
+	if err != nil {
+		config.Logger.Warn(err)
+		// only end the sensor process if we are in offline mode
+		if config.LidarDataRateMsec == 0 {
+			return strings.Contains(err.Error(), replaypcd.ErrEndOfDataset.Error())
```
Comment (PStrutz) [inline sensorprocess/sensorprocess.go:144]: [note to self] fix replaypcd line

File: sensorprocess/sensorprocess_test.go
Comment (PStrutz) [reply sensorprocess/sensorprocess_test.go:1]: **Testing Outline**

- rename some of the existing tests to Lidar, as they only cover lidar
- test tryAddIMUReadingUntilSuccess (offline mode)
- test tryAddIMUReadinfg (online mode) and check for correct time to sleep outputs
- test that adding IMU reading with undefined IMU errors
- test StartIMU, including reaching end of dataset
- no replay tests yet

File: viam-cartographer/src/carto_facade/carto_facade_test.cc
Comment (PStrutz) [inline viam-cartographer/src/carto_facade/carto_facade_test.cc:1]: **Testing Outline**

- test AddIMUReading, success and failure cases

File: cartofacade/carto_facade_test.go
Comment (PStrutz) [reply cartofacade/carto_facade_test.go:1]: **Testing Outline**

- test AddIMUReading, success and failure cases

File: cartofacade/carto_facade_mock.go
Comment (PStrutz) [inline cartofacade/carto_facade_mock.go:1]: **Testing outline**

- add support for a mock AddIMUReading 
- (to be used in integration testing)

File: cartofacade/capi_test.go
Comment (PStrutz) [reply cartofacade/capi_test.go:1]: **Testing Outline**

- test AddIMUReading wrapper
- test toIMUReading function
- create copy of TestCGoAPI to test both with only lidar and with lidar and imu, in the latter, add in IMU readings, including testing that invalid readings error
- already exists: test that movementsensor name is correctly converted in config

File: cartofacade/capi.go (hunk 1 above)
Comment (jeremyrhyde) [reply cartofacade/capi.go:80]: `r3.Vector` and `spatialmath.AngularVelocity`

File: cartofacade/carto_facade.go [hunk 9]
```diff
@@ -13,7 +13,7 @@ import (
 
 var emptyRequestParams = map[RequestParamType]interface{}{}
 
-// ErrUnableToAcquireLock is the error returned from AddSensorReading when lock can't be acquired.
+// ErrUnableToAcquireLock is the error returned from AddLidarReading when lock can't be acquired.
```
Comment (jeremyrhyde) [inline cartofacade/carto_facade.go:16]: [q] Should this be returned by `AddLidarReading` and `AddIMUReading`

File: cartofacade/carto_facade.go [hunk 10]
```diff
@@ -288,7 +319,24 @@ func (r *Request) doWork(
...
+		return nil, cf.carto.addLidarReading(lidar, reading, timestamp)
+	case addIMUReading:
+		imu, ok := r.requestParams[sensor].(string)
+		if !ok {
+			return nil, errors.New("could not cast inputted lidar name to string")
+		}
+
+		reading, ok := r.requestParams[reading].(IMUReading)
+		if !ok {
+			return nil, errors.New("could not cast inputted byte to byte slice")
```
Comment (jeremyrhyde) [inline cartofacade/carto_facade.go:331]: update error

File: cartofacade/carto_facade.go [hunk 11]
```diff
@@ -288,7 +319,24 @@ func (r *Request) doWork(
 			return nil, errors.New("could not cast inputted timestamp to times.Time")
 		}
 
-		return nil, cf.carto.addSensorReading(sensor, reading, timestamp)
+		return nil, cf.carto.addLidarReading(lidar, reading, timestamp)
+	case addIMUReading:
+		imu, ok := r.requestParams[sensor].(string)
+		if !ok {
+			return nil, errors.New("could not cast inputted lidar name to string")
```
Comment (jeremyrhyde) [inline cartofacade/carto_facade.go:326]: update error

File: sensorprocess/sensorprocess.go [hunk 12]
```diff
@@ -109,3 +112,96 @@ func tryAddSensorReading(ctx context.Context, reading []byte, readingTime time.T
...
+
+// addLidarReading adds a lidar reading to the cartofacade.
+func addIMUReading(
+	ctx context.Context,
+	config Config,
+) bool {
+	tsr, err := config.IMU.TimedIMUSensorReading(ctx)
+	if err != nil {
+		config.Logger.Warn(err)
+		// only end the sensor process if we are in offline mode
```
Comment (jeremyrhyde) [inline sensorprocess/sensorprocess.go:142]: [q] @nicksanford @PStrutz how do we want to determine offline mode for IMUs that are optional? Based on the Lidar data rate as seen below?

File: sensorprocess/sensorprocess.go [hunk 13]
```diff
@@ -109,3 +112,96 @@ func tryAddSensorReading(ctx context.Context, reading []byte, readingTime time.T
...
+	sr := cartofacade.IMUReading{
+		LinAccX: tsr.LinearAcceleration.X,
+		LinAccY: tsr.LinearAcceleration.Y,
+		LinAccZ: tsr.LinearAcceleration.Z,
+		AngVelX: tsr.AngularVelocity.X,
+		AngVelY: tsr.AngularVelocity.Y,
+		AngVelZ: tsr.AngularVelocity.Z,
+	}
+	/*
+	 when the lidar data rate msec is 0, we assume the user wants to be in "offline"
```
Comment (jeremyrhyde) [inline sensorprocess/sensorprocess.go:158]: update comment

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 14]
```diff
@@ -287,6 +287,7 @@ void CartoFacade::IOInit() {
             algo_config.missing_data_ray_length);
         map_builder.OverwriteMaxRange(algo_config.max_range);
         map_builder.OverwriteMinRange(algo_config.min_range);
+        map_builder.OverwriteUseIMUData(algo_config.use_imu_data);
```
Comment (jeremyrhyde) [inline viam-cartographer/src/carto_facade/carto_facade.cc:290]: [note to self] imu enabled must be defined for the `map_builder` AND during initialization in order to register the `kIMUSensor`

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 15]
```diff
@@ -741,38 +742,81 @@ void CartoFacade::Stop() {
...
     cartographer::transform::Rigid3d tmp_global_pose;
     bool update_latest_global_pose = false;
 
     if (map_builder_mutex.try_lock()) {
         VLOG(1) << "AddSensorData timestamp: " << measurement.time
+                << " Sensor type: Lidar "
                 << " measurement.ranges.size(): " << measurement.ranges.size();
-        map_builder.AddSensorData(measurement);
+        map_builder.AddSensorData(kRangeSensorId.id, measurement);
+        tmp_global_pose = map_builder.GetGlobalPose();
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply viam-cartographer/src/carto_facade/carto_facade.cc:777]: Concerns is too many mutex locks and unlocks even for the simple operation of copying pose

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 16]
```diff
@@ -1007,26 +1058,65 @@ extern int viam_carto_add_sensor_reading(viam_carto *vc,
...
-    rc = bdestroy(sr->sensor_reading);
+    // destroy lidar_reading
+    rc = bdestroy(sr->lidar_reading);
+    if (rc != BSTR_OK) {
+        return_code = VIAM_CARTO_DESTRUCTOR_ERROR;
+    }
+    sr->lidar_reading = nullptr;
+
+    // destroy sensor
+    rc = bdestroy(sr->lidar);
```
Comment (jeremyrhyde) [inline viam-cartographer/src/carto_facade/carto_facade.cc:1075]: [q] Same as above where was destroying lidar before this PR?

File: viam-cartographer/src/carto_facade/util.h [hunk 17]
```diff
@@ -49,9 +49,8 @@ void write_float_to_buffer_in_bytes(std::string &buffer, float f);
 
 void write_int_to_buffer_in_bytes(std::string &buffer, int d);
 
-std::tuple<bool, cartographer::sensor::TimedPointCloudData>
-carto_sensor_reading(std::string sensor_reading,
-                     int64_t sensor_reading_time_unix_milli);
+std::tuple<bool, cartographer::sensor::TimedPointCloudData> carto_lidar_reading(
```
Comment (jeremyrhyde) [inline viam-cartographer/src/carto_facade/util.h:52]: [q] Is there no associated `carto_imu_reading` function as the conversion is simple enough to be done inline?

File: viam-cartographer/src/carto_facade/map_builder.h [hunk 18]
```diff
@@ -72,8 +72,10 @@ class MapBuilder {
 
     // AddSensorData adds sensor data to cartographer's internal state
     // throws if adding sensor data fails.
-    // Currently assumes the sensor data is coming from the lidar
-    void AddSensorData(cartographer::sensor::TimedPointCloudData measurement);
+    void AddSensorData(const std::string &sensor_id,
```
Comment (jeremyrhyde) [inline viam-cartographer/src/carto_facade/map_builder.h:75]: [note] Different typing of these functions allow them to have the same name

File: viam-cartographer/src/carto_facade/map_builder.cc [hunk 19]
```diff
@@ -102,15 +102,28 @@ bool MapBuilder::SaveMapToFile(bool include_unfinished_submaps,
 }
 
 void MapBuilder::AddSensorData(
+    const std::string &sensor_id,
```
Comment (jeremyrhyde) [inline viam-cartographer/src/carto_facade/map_builder.cc:105]: [q] Why do we need sensor_id?

File: viam-cartographer/src/carto_facade/map_builder.cc [hunk 20]
```diff
@@ -102,15 +102,28 @@ bool MapBuilder::SaveMapToFile(bool include_unfinished_submaps,
...
-        {kRangeSensorId}, trajectory_builder_options_,
-        GetLocalSlamResultCallback());
+void MapBuilder::AddSensorData(const std::string &sensor_id,
+                               cartographer::sensor::ImuData measurement) {
+    trajectory_builder->AddSensorData(kIMUSensorId.id, measurement);
+}
+
+void MapBuilder::StartTrajectoryBuilder(bool use_imu_data) {
+    VLOG(1) << "MapBuilder::StartTrajectoryBuilder";
+    if (use_imu_data) {
```
Comment (jeremyrhyde) [inline viam-cartographer/src/carto_facade/map_builder.cc:117]: [nit] make a sensor list in if statement then one line 
```
trajectory_id = map_builder_->AddTrajectoryBuilder(
            sensorList, trajectory_builder_options_,
            GetLocalSlamResultCallback());
    }
```
outside of if statement

File: sensorprocess/sensorprocess_test.go
Comment (jeremyrhyde) [reply sensorprocess/sensorprocess_test.go:1]: Any edges cases to consider?

File: viam-cartographer/src/carto_facade/carto_facade_test.cc [hunk 21]
```diff
@@ -845,21 +912,675 @@ BOOST_AUTO_TEST_CASE(CartoFacade_demo) {
...
+        BOOST_TEST(viam_carto_add_lidar_reading(vc, &sr) == VIAM_CARTO_SUCCESS);
+        BOOST_TEST(viam_carto_add_lidar_reading_destroy(&sr) ==
+                   VIAM_CARTO_SUCCESS);
+    }
+
+    {
+        add_new_test_imu_reading(vc, 1629037853000020);
+        add_new_test_imu_reading(vc, 1629037853000040);
+        add_new_test_imu_reading(vc, 1629037853000060);
+        //add_new_test_imu_reading(vc, 1629037853000080);
```
Comment (jeremyrhyde) [inline viam-cartographer/src/carto_facade/carto_facade_test.cc:1397]: remove or uncomment comment

**Review by jeremyrhyde — COMMENTED** [review]
Initial review looks good. Will review again once all the tests are in. No major problems. Some refactoring of testing could be useful as test file as increased by 90%, can you look into combining cartofacade_demo and cartofacade_demo_with_imu and use a if statement when initializing/destorying the imu and adding readings?

@nicksanford As mentioned on slack, would love your review to focus on high-level logic, differences between imu and lidar as well as planned tests coverage. Let me or Patricia know if you have any large points/questions that would be good to go over in person before you leave on your trip.

File: sensorprocess/sensorprocess.go [hunk 22]
```diff
@@ -109,3 +112,96 @@ func tryAddSensorReading(ctx context.Context, reading []byte, readingTime time.T
...
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			err := config.CartoFacade.AddIMUReading(ctx, config.Timeout, config.IMUName, reading, readingTime)
+			if err == nil {
+				return
+			}
+			if !errors.Is(err, cartofacade.ErrUnableToAcquireLock) {
+				config.Logger.Warnw("Skipping sensor reading due to error from cartofacade", "error", err)
```
Comment (nicksanford) [inline sensorprocess/sensorprocess.go:187]: but this doesn't actually skip the reading as it is just going to retry again right? I think we should change this log to
```suggestion
				config.Logger.Warnw("Retrying sensor reading due to error from cartofacade", "error", err)
```

both here & in the lidar case.

File: sensorprocess/sensorprocess.go (hunk 12 above)
Comment (nicksanford) [reply sensorprocess/sensorprocess.go:142]: lets talk about this in team-slam-dev

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (nicksanford) [reply viam-cartographer/src/carto_facade/carto_facade.cc:777]: I agree its a concern. My vote would be to test how this code (which will give the most up to date pose, at the cost of lock contention) performs in practice & if we are dropping data too frequently due to lock contention we consider not updating the pose for IMU updates.

That test should happen before we consider the project complete.

I think we need to do that testing anyway, given we (or at least I) am not certain cartographer will give good outputs if the lidar & IMU readings are unsyncronized (as they are in this PR).

What do y'all think?

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 23]
```diff
@@ -930,10 +979,12 @@ extern int viam_carto_start(viam_carto *vc) {
         return VIAM_CARTO_VC_INVALID;
     }
     try {
+        std::cout << "this runs hahaha";
         viam::carto_facade::CartoFacade *cf =
             static_cast<viam::carto_facade::CartoFacade *>(vc->carto_obj);
         cf->Start();
     } catch (int err) {
+        std::cout << "damn";
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/carto_facade.cc:987]: please remove

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 24]
```diff
@@ -1007,26 +1058,65 @@ extern int viam_carto_add_sensor_reading(viam_carto *vc,
...
-        return VIAM_CARTO_SENSOR_READING_INVALID;
+        return VIAM_CARTO_LIDAR_READING_INVALID;
     }
     int return_code = VIAM_CARTO_SUCCESS;
     int rc = BSTR_OK;
-    // destroy sensor_reading
-    rc = bdestroy(sr->sensor_reading);
+    // destroy lidar_reading
+    rc = bdestroy(sr->lidar_reading);
+    if (rc != BSTR_OK) {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply viam-cartographer/src/carto_facade/carto_facade.cc:1069]: I think this was there before

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 25]
```diff
@@ -1007,26 +1058,65 @@ extern int viam_carto_add_sensor_reading(viam_carto *vc,
...
+extern int viam_carto_add_imu_reading_destroy(viam_carto_imu_reading *sr) {
+    if (sr == nullptr) {
+        return VIAM_CARTO_IMU_READING_INVALID;
+    }
+    int return_code = VIAM_CARTO_SUCCESS;
+    int rc = BSTR_OK;
 
     // destroy sensor
-    rc = bdestroy(sr->sensor);
+    rc = bdestroy(sr->imu);
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply viam-cartographer/src/carto_facade/carto_facade.cc:1115]: you shouldn't need to destroy the imu_readings if they are doubles

As long as you have C++ tests which test adding IMU readings & the tests pass valgrind tests (which run in CI) you shoulnd't be adding any memory leaks.

File: viam_cartographer.go [hunk 26]
```diff
@@ -122,21 +122,35 @@ func TerminateCartoLib() error {
 func initSensorProcess(cancelCtx context.Context, cartoSvc *CartographerService) {
```
Comment (nicksanford) [inline viam_cartographer.go:122]: please make plural as we now may be booting multiple

File: viam_cartographer.go [hunk 27]
```diff
@@ -122,21 +122,35 @@ func TerminateCartoLib() error {
...
 	go func() {
 		defer cartoSvc.sensorProcessWorkers.Done()
-		if jobDone := sensorprocess.Start(cancelCtx, spConfig); jobDone {
+		if jobDone := sensorprocess.StartLidar(cancelCtx, spConfig); jobDone {
 			cartoSvc.jobDone.Store(true)
 			cartoSvc.cancelSensorProcessFunc()
 		}
 	}()
+
+	if spConfig.IMUName != "" {
```
Comment (nicksanford) [inline viam_cartographer.go:144]: What happens if IMUName is non empty but `config_params.use_imu_data` is true? or vice versa?

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply viam-cartographer/src/carto_facade/carto_facade.cc:777]: Talking in team-slam-dev, will post resolution here once consensus is achieved.

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 25 above)
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply viam-cartographer/src/carto_facade/carto_facade.cc:1115]: Sorry, that comment was meant to be dont need to destroy them

File: sensorprocess/sensorprocess.go (hunk 12 above)
Comment (jeremyrhyde) [reply sensorprocess/sensorprocess.go:142]: Determine based on lidar data rate, if lidar data rate is 0 and imu is not, error

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply viam-cartographer/src/carto_facade/carto_facade.cc:777]: Steps to test:
1. Run in offline and download logs for:
- Only lidar
- Lidar + imu and constant updating pose
- Lidar + imu and only updating pose when lidar data is added
2. Grep log files (ctrl-F should be fine to) searching for drop logs
3. Compare number of drops

File: cartofacade/capi.go (hunk 1 above)
Comment (PStrutz) [reply cartofacade/capi.go:80]: Done

File: cartofacade/capi.go [hunk 28]
```diff
@@ -203,17 +215,35 @@ func (vc *Carto) terminate() error {
 	return nil
 }
 
-// AddSensorReading is a wrapper for viam_carto_add_sensor_reading
-func (vc *Carto) addSensorReading(sensor string, readings []byte, timestamp time.Time) error {
-	value := toSensorReading(sensor, readings, timestamp)
+// AddLidarReading is a wrapper for viam_carto_add_lidar_reading
```
[the remark this answers is not in the document]
Comment (PStrutz) [reply cartofacade/capi.go:218]: Done

File: cartofacade/capi.go [hunk 29]
```diff
@@ -203,17 +215,35 @@ func (vc *Carto) terminate() error {
...
-	status = C.viam_carto_add_sensor_reading_destroy(&value)
+	status = C.viam_carto_add_lidar_reading_destroy(&value)
+	if err := toError(status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AddIMUReading is a wrapper for viam_carto_add_imu_reading
```
[the remark this answers is not in the document]
Comment (PStrutz) [reply cartofacade/capi.go:236]: Done

File: cartofacade/capi.go (hunk 3 above)
Comment (PStrutz) [reply cartofacade/capi.go:504]: Done

File: cartofacade/capi.go (hunk 2 above)
Comment (PStrutz) [reply cartofacade/capi.go:506]: Done

File: sensorprocess/sensorprocess.go (hunk 22 above)
Comment (PStrutz) [reply sensorprocess/sensorprocess.go:187]: Changed it for both.

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 14 above)
Comment (PStrutz) [reply viam-cartographer/src/carto_facade/carto_facade.cc:290]: See Viam-cartographer.go lines 440-445

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 30]
```diff
@@ -741,38 +742,81 @@ void CartoFacade::Stop() {
...
+    measurement.time =
+        cartographer::common::FromUniversal(0) +
+        cartographer::common::FromMilliseconds(imu_reading_time_unix_milli);
+    measurement.linear_acceleration =
+        Eigen::Vector3d(sr->lin_acc_x, sr->lin_acc_y, sr->lin_acc_z);
+    measurement.angular_velocity =
+        Eigen::Vector3d(sr->ang_vel_x, sr->ang_vel_y, sr->ang_vel_z);
+
+    cartographer::transform::Rigid3d tmp_global_pose;
+    bool update_latest_global_pose = false;
```
[the remark this answers is not in the document]
Comment (PStrutz) [reply viam-cartographer/src/carto_facade/carto_facade.cc:814]: Not sure - it's also on line 770 for the lidar function

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 24 above)
[the remark this answers is not in the document]
Comment (PStrutz) [reply viam-cartographer/src/carto_facade/carto_facade.cc:1069]: I agree with Nick, this is definitely not something I just pulled out of thin air. No idea why git isn't showing the change correctly

File: viam-cartographer/src/carto_facade/map_builder.h (hunk 18 above)
Comment (PStrutz) [reply viam-cartographer/src/carto_facade/map_builder.h:75]: Yep, this is what sensor_id is needed for

File: viam-cartographer/src/carto_facade/util.h (hunk 17 above)
Comment (PStrutz) [reply viam-cartographer/src/carto_facade/util.h:52]: Yes

File: viam_cartographer.go (hunk 27 above)
Comment (PStrutz) [reply viam_cartographer.go:144]: I am slightly confused - if IMU name is nonempty, then we want to be using emu data, so the variable should be true, right?
If you are asking about the case in which the variable is set to true, but we do not have an IMU configured, or we have an IMU configured, but the variable is set to false - this is addressed in my upcoming commit, see lines 440-445 of viam_cartographer.go

File: cartofacade/capi_test.go
Comment (PStrutz) [reply cartofacade/capi_test.go:1]: Everything done, except that adding IMU sensor readings fails to create nonempty point cloud map

File: cartofacade/carto_facade_test.go
Comment (jeremyrhyde) [reply cartofacade/carto_facade_test.go:1]: Can we resolve?

File: integration_test.go [hunk 31]
```diff
@@ -51,15 +51,28 @@ func testCartographerMap(t *testing.T, svc slam.Service, localizationMode bool)
...
+	var expectedOriLinux *spatialmath.R4AA
 	tolerancePos := 0.001
-
-	expectedOriOSX := &spatialmath.R4AA{Theta: 1.5465081272043815, RX: 0, RY: 0, RZ: 1}
-	expectedOriLinux := &spatialmath.R4AA{Theta: 0.3331667853231311, RX: 0, RY: 0, RZ: 1}
 	toleranceOri := 0.001
 
+	if useIMU {
+		expectedPosOSX = r3.Vector{X: 31.26644008021215, Y: -0.07725723487584407, Z: 0}
+		expectedPosLinux = r3.Vector{X: 158.79903385710674, Y: -77.01514065531592, Z: 0}
```
Comment (jeremyrhyde) [inline integration_test.go:64]: update values for linux + imu

File: integration_test.go [hunk 32]
```diff
@@ -158,15 +172,14 @@ func testHelperCartographer(
 		test.That(t, errors.New("test timeout"), test.ShouldBeNil)
```
Comment (jeremyrhyde) [inline integration_test.go:172]: [minor] add comment here saying we will check both channels once a accurate mock dataset has been gathered and like the JIRA ticket for that

File: integration_test.go [hunk 33]
```diff
@@ -261,34 +263,34 @@ func integrationtestHelperCartographer(t *testing.T, subAlgo viamcartographer.Su
 		// save the internal state of the mapping run to a new datadir
 		saveInternalState(t, internalState, dataDirectoryLocalizing)
 		// localize on that internal state
-		testHelperCartographer(t, dataDirectoryLocalizing, subAlgo, logger, false, true, 0, cartofacade.LocalizingMode)
+		testHelperCartographer(t, dataDirectoryLocalizing, subAlgo, logger, true, false, 0, cartofacade.LocalizingMode)
 	})
 
-	t.Run("replay sensor localizing mode", func(t *testing.T) {
```
Comment (jeremyrhyde) [inline integration_test.go:267]: [q] did these tests switch orders? I see a similar change on line 245

File: sensorprocess/sensorprocess.go (hunk 13 above)
Comment (jeremyrhyde) [reply sensorprocess/sensorprocess.go:159]: I think my comment here was in regard to using the imu data rate not lidar but that was resolved and comment remains correct

File: sensors/sensors.go [hunk 34]
```diff
@@ -72,16 +72,17 @@ func NewLidar(
 		return Lidar{}, errors.Wrapf(err, "error getting lidar camera %v for slam service", cameraName)
 	}
 
-	// If there is a camera provided in the 'camera' field, we enforce that it supports PCD.
-	properties, err := newLidar.Properties(ctx)
-	if err != nil {
-		return Lidar{}, errors.Wrapf(err, "error getting lidar camera properties %v for slam service", cameraName)
-	}
+	// rplidar does not yet support properties, see https://viam.atlassian.net/browse/RSDK-4600
```
Comment (jeremyrhyde) [inline sensors/sensors.go:75]: hot fix is in, you should be able to comment back in this code

File: sensors/sensors_test.go [hunk 35]
```diff
@@ -23,9 +23,10 @@ func TestValidateGetLidarData(t *testing.T) {
 	goodLidar, err := s.NewLidar(ctx, s.SetupDeps(lidar, ""), lidar, logger)
 	test.That(t, err, test.ShouldBeNil)
 
-	lidar = "lidar_with_invalid_properties"
-	_, err = s.NewLidar(ctx, s.SetupDeps(lidar, ""), lidar, logger)
-	test.That(t, err, test.ShouldBeError, errors.New("configuring lidar camera error: 'camera' must support PCD"))
+	// Implement once rplidar supports Properties, see https://viam.atlassian.net/browse/RSDK-4600
+	// lidar = "lidar_with_invalid_properties"
```
Comment (jeremyrhyde) [inline sensors/sensors_test.go:27]: same here

File: sensorprocess/sensorprocess_test.go
Comment (PStrutz) [reply sensorprocess/sensorprocess_test.go:1]: Yes I think so

File: cartofacade/capi_test.go
Comment (PStrutz) [reply cartofacade/capi_test.go:1]: Yes

File: cartofacade/carto_facade_test.go
Comment (PStrutz) [reply cartofacade/carto_facade_test.go:1]: Yes

File: viam_cartographer.go (hunk 27 above)
Comment (PStrutz) [reply viam_cartographer.go:145]: Wouldn't it be good to log some sort of indication that it is being configured to use an IMU vs not?

File: integration_test.go (hunk 32 above)
Comment (PStrutz) [reply integration_test.go:172]: Done

File: integration_test.go (hunk 33 above)
Comment (PStrutz) [reply integration_test.go:267]: No they did not, GitHub marking these lines might have to do with changing the use imu / don't use imu test ordering

File: integration_test.go [hunk 36]
```diff
@@ -261,34 +263,34 @@ func integrationtestHelperCartographer(t *testing.T, subAlgo viamcartographer.Su
...
+			err := os.RemoveAll(dataDirectoryUpdating)
 			test.That(t, err, test.ShouldBeNil)
 		}()
 
 		// save the internal state of the mapping run to a new datadir
-		saveInternalState(t, internalState, dataDirectoryLocalizing)
-		// localize on that internal state
-		testHelperCartographer(t, dataDirectoryLocalizing, subAlgo, logger, true, false, 0, cartofacade.LocalizingMode)
+		saveInternalState(t, internalState, dataDirectoryUpdating)
+		// update fromthat internal state
```
[the remark this answers is not in the document]
Comment (PStrutz) [reply integration_test.go:289]: Done

File: sensors/sensors.go (hunk 34 above)
Comment (PStrutz) [reply sensors/sensors.go:75]: Yes and no - I'm having issues installing rplidar module from source

File: testhelper/testhelper.go [hunk 37]
```diff
@@ -267,7 +267,7 @@ func IntegrationTimedIMUSensor(
 		if int(i) >= len(mockLinearAccelerationData) {
 			// communicate to the test that all imu readings have been written
 			if !closed {
-				done <- struct{}{}
+				// done <- struct{}{}
```
[the remark this answers is not in the document]
Comment (PStrutz) [reply testhelper/testhelper.go:270]: Done

File: integration_test.go (hunk 33 above)
Comment (jeremyrhyde) [reply integration_test.go:267]: Probably the linter more than github

File: sensors/sensors.go (hunk 34 above)
Comment (jeremyrhyde) [reply sensors/sensors.go:75]: Spoke about offline

**Pull request closed without merging**